partiql-planner/src/main/kotlin/org/partiql/planner/internal/transforms/RexConverter.kt

```diff
@@ -249,41 +259,106 @@ internal object RexConverter {
+                    relBinding(
+                        name = "_k$index", // fresh variable
+                        type = StaticType.STRING
+                    ),
+                    relBinding(
+                        name = "_v$index", // fresh variable
+                        type = path.type
+                    )
+                )
+                val props = emptySet<Rel.Prop>()
+                val relType = relType(schema, props)
+                rel(relType, relOpUnpivot(path))
+            }
+
+            val fromList = mutableListOf<Rel>()
+
+            val pathNavi = newSteps.fold(newRoot) { current, step ->
+                val path = when (step) {
+                    is Expr.Path.Step.Index -> {
+                        val key = visitExprCoerce(step.key, context)
+                        val op = when (val astKey = step.key) {
+                            is Expr.Lit -> when (astKey.value) {
+                                is StringValue -> rexOpPathKey(current, key)
                                 else -> rexOpPathIndex(current, key)
                             }
-                        }
-                        is Expr.Path.Step.Symbol -> {
-                            val identifier = AstToPlan.convert(step.symbol)
-                            when (identifier.caseSensitivity) {
-                                Identifier.CaseSensitivity.SENSITIVE -> rexOpPathKey(
-                                    current,
-                                    rexString(identifier.symbol)
-                                )
-                                Identifier.CaseSensitivity.INSENSITIVE -> rexOpPathSymbol(current, identifier.symbol)
+
+                            is Expr.Cast -> when (astKey.asType is Type.String) {
+                                true -> rexOpPathKey(current, key)
+                                false -> rexOpPathIndex(current, key)
                             }
+
+                            else -> rexOpPathIndex(current, key)
                         }
-                        is Expr.Path.Step.Unpivot -> error("Unpivot path not supported yet")
-                        is Expr.Path.Step.Wildcard -> error("Wildcard path not supported yet")
+                        op
+                    }
+
+                    is Expr.Path.Step.Symbol -> {
+                        val identifier = AstToPlan.convert(step.symbol)
+                        val op = when (identifier.caseSensitivity) {
+                            Identifier.CaseSensitivity.SENSITIVE -> rexOpPathKey(
+                                current,
+                                rexString(identifier.symbol)
+                            )
+
+                            Identifier.CaseSensitivity.INSENSITIVE -> rexOpPathSymbol(current, identifier.symbol)
+                        }
+                        op
+                    }
+
+                    is Expr.Path.Step.Unpivot -> {
+                        val op = rexOpVarLocal(1, -1)
```

Comment:
While this works, I think it may be subject to breakage, especially if we don't document why `-1` is being used here (I see its use in PlanTyper). Would it just be better to calculate the `ref` here? Seemingly, it's derived from the `fromList` schemas.

Reply:
The `rexOpVarLocal(1,-1)` is used to indicate that the var is coming from the last element from the immediate outer scope. I think this is the correct modeling in this rewrite. 

From the spec: 

```
Consider the path collection expression:
ew1p1 . . . wnpn
where e is any expression; n > 0; each wildcard step wi is either [∗] or .∗; each series of plain path steps pi is a sequence of zero or more tuple path navigations or array navigations (potentially mixed).

Then the path collection expression is equivalent to : 
SELECT VALUE vn.pn
FROM
u1 e AS v1,
u2 @v1.p1 AS v2,
. . .,
un @vn−1.pn−1 AS vn,

where each vi is a fresh variable and each ui is UNPIVOT if wi is a .∗ and it is nothing if wi is a [∗]. Intuitively vi corresponds to the i-th star.
```

So this (relatively massive fold), would continue folding on non wild card steps, and upon encounter wildcard step, it would rewrite the tip of the fold to be the last var in the immediate outer scope.  (`rexOpVar(1,-1)`)

Once folding is completed, we check if the fromList is empty, if not, then we must rewrite. 
The from then is composite with one or more scan node (in case of more than one scan, they are combined using inner join). 

The local scope of the select construct can see all the bindings produced by each of the scan node, the one that it actually need to use the bonded to be the last one, i.e., `rexOpVar(0,-1)`. 

```
input: a.b.c 
  |_ current: a
     |_ b <- Not a wild card, current a.b 
       |_ c <- not a wild card, current a.b.c 

After fold: return a.b.c as a rex
```

```
input a[*]
   |_ current : a 
      |_ wild card:  add scan(a) to from list, current var(1, -1)

Note in this step, it should be identical to say this is var(1,0).

After fold: the op of path navi would be rexOpVar(1,-1)

We construct the select constructor to essentially rewrite this to be rexOpVar(0,-1), the last var in the local scope. 

That is equivalent to:

SELECT VALUE v1
FROM a as v1
```

```
input a[*].b
   |_ current : a 
      |_ wild card:  add scan(a) to from list, current var(1, -1)
         |_ b : var(1,-1).b

After fold: the op of path navi would be var(1,-1).b

We construct the select constructor to essentially rewrite this to be Var(0,-1).b, the last var in the local scope. 

That is equivalent to:

SELECT VALUE v1.b
FROM a as v1
```

```
input a.*
   |_ current: a 
     |_ unpivot: add unpivot(a) to from list, current var(1,-1)

Note that unpivot is interesting in terms of that it creates two bindings in the scope: key and value, where key is the first and value is the second. we use the -1 to points to second var. So in reality this is the same as var(1,1). 

That is equivalent to:

SELECT VALUE v1
FROM UNPIVOT a as v1
```

```
input a.*b
   |_ current: a 
     |_ unpivot: add unpivot(a) to from list, current var(1,-1)
       |_ b -> current var(1,-1).b

Notice that in this case, the from list contains 1 node, but because the node is unpivot, the number of bindings the select constructs holds locally should be two, So we currently when constructing var(0,-1) in the select constructor, the var would be resolved to var(0, 1) in this case instead of var(0,0). 

That is equivalent to:

SELECT VALUE v1.b
FROM UNPIVOT a as v1
```

Reply:
Summary: 
For individual FROM node: we can resolve the ref here, but I choose `-1` to model directly on `the last element of immediate out scope`. 

For select construct: we can also resolve the ref here, but the code is a little more completed as you would have to calculate the number of scan and unpivot to give the current index. This is why I use the -1, I believe it is always going to be the last var.

Reply:
We discussed offline, and I'm leaving some details.

The use of `-1` reminded me of [this use of negative indices](https://github.com/partiql/partiql-lang-kotlin/blob/1c7dce11cc9fad417e5ea97e1ec6df0283174334/partiql-planner/src/main/kotlin/org/partiql/planner/internal/transforms/PlanTransform.kt#L370). Since the contract of using negative indices is potentially weak and since we have access to the input schema, we can calculate the index at this stage.